### PR TITLE
[java] Support JDK 18

### LINKED ISSUE #3809: [java] Support JDK 18
[OpenJDK 18](https://openjdk.java.net/projects/jdk/18/)  GA is scheduled for 2022/03/22.
So we should add support for that with the next PMD version.

The following JEPs are important for us:
* [JEP 413](https://openjdk.java.net/jeps/413): Code Snippets in Java API Documentation
* [JEP 420](https://openjdk.java.net/jeps/420): Pattern Matching for switch (Second Preview)

In summary: https://javaalmanac.io/jdk/18/

This issue is similar to the previous: #3366

EA Release of JDK18 is available here: http://jdk.java.net/18/

**Tasks:**

- [x] Add new java language versions 18 and 18-preview.
    - 18 is the new default
    - update also java.md and ant.md
    - and add in KotlinTestingDsl.kt.
- [x] Update asm dependency:
    - https://search.maven.org/artifact/org.ow2.asm/asm
    - https://asm.ow2.io/versions.html
    - Update net.sourceforge.pmd.lang.java.typeresolution.visitors.PMDASMVisitor.ASM_API if necessary
    - :heavy_check_mark: we already use the latest version 9.2
- [x] Update implementation for "Pattern Matching for switch" (406, 420) as a preview feature, mark AST classes as experimental, if needed.
    - https://openjdk.java.net/jeps/406 (preview 1)
    - https://openjdk.java.net/jeps/420 (preview 2)
    - http://cr.openjdk.java.net/~gbierman/jep420/jep420-20211208/specs/patterns-switch-jls.html
- [x] Remove java language version 16-preview
    - update also java.md and ant.md
    - update KotlinTestingDsl.kt
- [x] Write release notes
- [x] Prepare merge for PMD 7 -> #3838

### PR DESCRIPTION
**Pull request by adangel**
- Fixes #3809 

- Note - Java 18 has no new standard language features, so I didn't create a Java18TreeDumpTest.
- I didn't find any language change in the preview features for pattern matching, so basically no grammar changes are done (except for the version numbers).
- I enhanced UnnecessaryImportRule to support the new `@snippet // @link` syntax in javadoc (JEP 413).

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java [hunk 1]
```diff
@@ -1,9 +0,0 @@
-/*
- * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
- */
-package com.example.expression;
-
-/**
- * @see <a href="https://openjdk.java.net/jeps/397">JEP 397: Sealed Classes (Second Preview)</a>
- */
-public final class ConstantExpr implements Expr { }
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java:9]: Why don't we keep these tests, if the features are identical? Where do the new files for java18p come?

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml [hunk 2]
```diff
@@ -912,6 +912,29 @@ public class U {
...
+
+    <test-code>
+        <description>Necessary imports for @snippet tags introduced with JEP 413 in Java 18</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+import java.math.BigDecimal;
+import java.util.ArrayList;
+
+/**
+ * {@snippet :
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:927]: Da muss ich jetzt meinen Javadoc-parser aktualisieren...

File: docs/pages/pmd/languages/java.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+---
+title: Supported Java Versions
+permalink: pmd_languages_java.html
```
[the remark this answers is not in the document]
Comment (adangel) [reply docs/pages/pmd/languages/java.md:3]: I've kept in as a separate file in pmd7 (docs/pages/pmd/languages/java_versions.md, #3838).
I'm not sure, what the best approach is. The permalink will change from pmd_languages_java.html to pmd_languages_java_versions.html ...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java (hunk 1 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java:9]: Maybe approach is this: delete the old preview tests (in that case 16-preview) without looking at it at all. If the preview features was standardized, then we should have a test for e.g. 17. In that case, these files are really in 17 already (https://github.com/pmd/pmd/blob/master/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java17/expression/ConstantExpr.java). 

When adding a new language preview version, I copy the previous test files. For this update, I copied 17-preview into 18-preview and reviewed the test cases. E.g. I added a bit more code samples from the new JEP. Luckily, there were no grammar changes. I didn't touch 17-preview at all. But 18-preview contains all cases from 17-preview plus some more.

So rule of thumb: whenever a preview feature is standardized, we need to take over the test cases from preview (maybe with additional changes).

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml (hunk 2 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/UnnecessaryImport.xml:927]: btw. I didn't try this with the javadoc tooling. I hope, I understood the specification in the JEP correctly... :)

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java (hunk 1 above)
Comment (adangel) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/ast/jdkversiontests/java16p/expression/ConstantExpr.java:9]: Hm... we might think about whether to keep 17-preview when we add 19 in September later this year. 17 is a LTS release, so potentially someone could need it. The only preview feature from 17-preview is pattern matching for switch, which would be standardized in 19 though (probably).

**Pull request merged**